edgedb-protocol/src/query_arg.rs

```diff
@@ -526,3 +527,99 @@ implement_tuple! {9, T0, T1, T2, T3, T4, T5, T6, T7, T8, }
 implement_tuple! {10, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, }
 implement_tuple! {11, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, }
 implement_tuple! {12, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, }
+
+
+
+pub struct ValueWithCardinality(Option<Value>, Cardinality);
+
+impl<V: Into<Value>> From<V> for ValueWithCardinality {
+	fn from(value: V) -> Self {
+		ValueWithCardinality(Some(value.into()), Cardinality::One)
+	}
+}
+
+impl<V: Into<Value>> From<Option<V>> for ValueWithCardinality
+    where Value: From<V> 
+{
+	fn from(value: Option<V>) -> Self {
+		ValueWithCardinality(value.map(Value::from), Cardinality::AtMostOne)
+	}
+}
+
+impl<V: Into<Value>> From<Vec<V>> for ValueWithCardinality
+    where Value: From<V>
+{
+	fn from(value: Vec<V>) -> Self {
+		ValueWithCardinality(
+			Some(Value::Array(value.into_iter().map(Value::from).collect())),
+			Cardinality::One
+		)
+	}
+}
+
+impl<V: Into<Value>> From<Option<Vec<V>>> for ValueWithCardinality
+    where Value: From<V>
+{
+	fn from(value: Option<Vec<V>>) -> Self {
+		let mapped = value.map(|value| Value::Array(value.into_iter().map(Value::from).collect()));
+		ValueWithCardinality(mapped, Cardinality::AtMostOne)
+	}
+}
+
+pub fn value_from_pairs<K>(pairs: &HashMap<K, ValueWithCardinality>) -> Value
+where
+	K: ToString,
+{
+	let mut elements = Vec::new();
+	let mut fields: Vec<Option<Value>> = Vec::new();
+	for (key, arg) in pairs.iter() {
+		let ValueWithCardinality(value, cd) = arg;
+
+		elements.push(ShapeElement {
+			name: key.to_string(),
+			cardinality: Some(*cd),
+			flag_link: false,
+			flag_link_property: false,
+			flag_implicit: false
+		});
+		fields.push(value.clone());
```

Comment:
I don't like .clone() here, I think it's possible to avoid it, but I'm not sure how!